### PR TITLE
PersistenceLengths and its unitary tests

### PR DESCRIPTION
**Pull request by VincentRouvreau**

### CODE REVIEW

File: src/python/gudhi/representations/vector_methods.py [hunk 1]
```diff
@@ -885,3 +885,64 @@ def transform(self, X, sample_weight=None):
...
+            X (list of n x 2 numpy arrays): input persistence diagrams.
+    
+        Returns:
+            numpy array with shape (number of diagrams) x (num_lengths): output persistence lengths.
+        """
+
+        pers_lengths = []
+        for pd in X:
+            # Sort in reverse order persistence lengths (where length = death - birth)
+            lengths = np.flip(np.sort(pd[:,1] - pd[:,0]))
```
Comment (mglisse) [inline src/python/gudhi/representations/vector_methods.py:927]: To get the top k elements (with k significantly smaller than n), a slightly more efficient algorithm is to use partition(*, -k)[-k:] (all the numbers I give are ±1, to be checked) and then you only need to sort those k elements. But that can come later if you prefer.

Comment (VincentRouvreau) [reply src/python/gudhi/representations/vector_methods.py]: I did so on [768774a](https://github.com/GUDHI/gudhi-devel/pull/1117/commits/768774a184139fcddd168d0ee6e5e4f5ce8eaa15), but I had to assert on `num_lengths <= 0` in the constructor. To be discussed, cf. [this discussion in Ripser PR](https://github.com/GUDHI/gudhi-devel/pull/920#discussion_r1741115600)

File: src/python/gudhi/representations/vector_methods.py [hunk 2]
```diff
@@ -912,28 +918,29 @@ def fit(self, X, y=None):
 
     def transform(self, X):
         """
-        Compute the persistence landscape for each persistence diagram individually and concatenate the results.
+        Compute the persistence lengths for each persistence diagram individually and concatenate the results.
```
Comment (mglisse) [inline src/python/gudhi/representations/vector_methods.py:921]: When I read "concatenate", I expect the output to be a 1d array of length N*k.

File: src/python/gudhi/representations/vector_methods.py [hunk 3]
```diff
@@ -912,28 +918,29 @@ def fit(self, X, y=None):
...
 
         Parameters:
             X (list of n x 2 numpy arrays): input persistence diagrams.
-    
+
         Returns:
             numpy array with shape (number of diagrams) x (num_lengths): output persistence lengths.
```
Comment (mglisse) [inline src/python/gudhi/representations/vector_methods.py:927]: That would be nice, but isn't the current output a list?

File: src/python/gudhi/representations/vector_methods.py (hunk 2 above)
Comment (VincentRouvreau) [reply src/python/gudhi/representations/vector_methods.py]: I removed the "concatenate" on [3bbd3af](https://github.com/GUDHI/gudhi-devel/pull/1117/commits/3bbd3af834902e79e83b990ae246dc9f1469e2d7)

File: src/python/gudhi/representations/vector_methods.py (hunk 3 above)
Comment (VincentRouvreau) [reply src/python/gudhi/representations/vector_methods.py:927]: You are right, so I took the opportunity to rewrite it with a first result array filled with zeros, that is instantiated rows by rows on [3bbd3af](https://github.com/GUDHI/gudhi-devel/pull/1117/commits/3bbd3af834902e79e83b990ae246dc9f1469e2d7)

File: src/python/gudhi/representations/vector_methods.py [hunk 4]
```diff
@@ -885,3 +885,71 @@ def transform(self, X, sample_weight=None):
...
+
+        Parameters:
+            X (list of n x 2 numpy arrays): input persistence diagrams.
+
+        Returns:
+            numpy array with shape (number of diagrams) x (num_lengths): output persistence lengths.
+        """
+        pers_length_array = np.zeros((len(X), self.num_lengths))
+        idx = 0
+        for pd in X:
```
[the remark this answers is not in the document]
Comment (VincentRouvreau) [reply src/python/gudhi/representations/vector_methods.py]: Oh, nice ! I did it on [af41754](https://github.com/GUDHI/gudhi-devel/pull/1117/commits/af4175427275a7fc82632d657962b3093c402fba)

File: src/python/gudhi/representations/vector_methods.py [hunk 5]
```diff
@@ -885,3 +885,69 @@ def transform(self, X, sample_weight=None):
...
+            raise ValueError("num_lengths must be greater than 0.")
+        self.num_lengths = num_lengths
+
+    def fit(self, X, y=None):
+        """
+        Fit the PersistenceLengths class on a list of persistence diagrams (this function actually does nothing but is
+        useful when PersistenceLengths is included in a scikit-learn Pipeline).
+
+        Parameters:
+            X (list of n x 2 or n x 1 numpy arrays): input persistence diagrams.
```
Comment (mglisse) [inline src/python/gudhi/representations/vector_methods.py:914]: Why nx1?

File: src/python/gudhi/representations/vector_methods.py [hunk 6]
```diff
@@ -885,3 +885,69 @@ def transform(self, X, sample_weight=None):
...
+            y (n x 1 array): persistence diagram lengths (unused).
```
[the remark this answers is not in the document]
Comment (VincentRouvreau) [reply src/python/gudhi/representations/vector_methods.py]: Sounds good to me. I changed it on [009748c](https://github.com/GUDHI/gudhi-devel/pull/1117/commits/009748ca5b3ad81712c33bfcd169239c5a4f0cd5)

File: src/python/gudhi/representations/vector_methods.py (hunk 5 above)
Comment (VincentRouvreau) [reply src/python/gudhi/representations/vector_methods.py]: Thanks, I missed this copy /paste that has to be changed. I fixed it on [e984496](https://github.com/GUDHI/gudhi-devel/pull/1117/commits/e984496218c0ee4b463a57e3ca6ceb2ffabe1945)

**Review by mglisse — APPROVED** [review]
Looks ok to me, give @MathieuCarriere a bit of time to comment.

**Pull request merged**